sandboxed-fs.js

```diff
@@ -4,11 +4,109 @@ const sandboxedFs = {};
 module.exports = sandboxedFs;
 
 const fs = require('fs');
+const pathModule = require('path');
+
+const errorMessage = 'path must be a string';
+
+function makePathSafe(path) {
+  return pathModule.resolve('/', path);
+}
+
+const pathFunctionsWrapper = (func, path) => (p, ...args) => {
+  if (typeof p === 'string') {
+    return func(pathModule.join(path, makePathSafe(p)), ...args);
+  }
+
+  throw new TypeError(errorMessage);
+};
+
+const fileFunctionsWrapper = (func, path) => (file, ...args) => {
+  if (typeof file === 'string') {
+    return func(pathModule.join(path, makePathSafe(file)), ...args);
+  }
+  if (typeof file === 'number') {
+    return func(p, ...args);
+  }
+
+  throw new TypeError(errorMessage);
+};
+
+const twoPathFunctionsWrapper = (func, path) => (p1, p2, ...args) => {
+  if (typeof p1 === 'string' || typeof p2 === 'string') {
+    return func(pathModule.join(path, makePathSafe(p1)),
+                pathModule.join(path, makePathSafe(p2)), ...args);
+  }
+
+  throw new TypeError(errorMessage);
+};
+
+const functionTypes = {
+  pathFunctions: {
+    names: [
+      'access',
+      'chmod',
+      'chown',
+      'exists',
+      'lstat',
+      'mkdir',
+      'mkdtemp',
+      'open',
+      'readdir',
+      'readlink',
+      'realpath',
+      'rmdir',
+      'stat',
+      'truncate',
+      'unlink',
+      'utimes'
+    ],
+    wrapper: pathFunctionsWrapper,
+    isSync: true
+  },
+  pathNonSyncFunctions: {
+    names: [
+      'createReadStream',
+      'createWriteStream',
+      'unwatchFile',
+      'watch',
+      'watchFile'
+    ],
+    wrapper: pathFunctionsWrapper,
+    isSync: false
+  },
+  fileFunctions: {
+    names: [
+      'appendFile',
+      'readFile',
+      'writeFile'
+    ],
+    wrapper: fileFunctionsWrapper,
+    isSync: true
+  },
+  twoPathFunctions: {
+    names: [
+      'link',
+      'rename',
+      'symlink'
+    ],
+    wrapper: twoPathFunctionsWrapper,
+    isSync: true
+  }
+};
 
 sandboxedFs.bind = (path) => {
-  const wrapped = fs;
+  const wrapped = Object.assign({}, fs);
 
-  // wrapper implementation
+  for (const typeName in functionTypes) {
```

Comment:
I'd prefer not to use `for-in` loops at all. OTOH it's not a hot code path, so it isn't that critical.